server/src/services/sqlIngest.py

```diff
@@ -1,254 +1,148 @@
-                connection.execute(delete_stmt)
-                updated += 1
-            else:
-                inserted += 1
-            # Write record
-            insert_stmt = db.insert(staging)
-            row = self.data[self.data.srnumber == srnumber].to_dict('results')
-            row = [fix_nan_vals(r) for r in row]
-            connection.execute(insert_stmt, row)
-        print('Operation Complete: %d inserts, %d updates in %.2f minutes' %
-              (inserted, updated, self.elapsedTimer(updateTimer)))
-
-
-if __name__ == "__main__":
-    '''Class DataHandler workflow from initial load to SQL population'''
-    config = ConfigParser()
-    config.read('../settings.cfg')
-    loader = DataHandler(config)
-    loader.fetchSocrataFull()
-    loader.cleanData()
-    loader.ingestData(tableName='ingest_staging_table')
+
+        minutes = timer.end()
+        log('\tDone with {} after {} minutes.'.format(year, minutes))
+        log('\tRows inserted: {}'.format(rowsInserted))
+
+        return {
+            'year': year,
+            'rowsInserted': rowsInserted,
+            'endReached': endReached,
+            'minutesElapsed': minutes,
+        }
+
+    def cleanTable(self):
+        def exec_sql(sql):
+            with self.engine.connect() as conn:
+                return conn.execute(text(sql))
+
+        def dropDuplicates(table, report):
+            rows = exec_sql(f"""
+                DELETE FROM {table} a USING {table} b
+                WHERE a.id < b.id AND a.srnumber = b.srnumber;
+            """)
+
+            report.append({
+                'description': 'dropped duplicate rows by srnumber',
+                'rows': rows.rowcount
+            })
+
+        def switchPrimaryKey(table, report):
+            exec_sql(f"""
+                ALTER TABLE {table} DROP COLUMN id;
+                ALTER TABLE {table} ADD PRIMARY KEY (srnumber);
+            """)
+
+            report.append({
+                'description': 'switched primary key column to srnumber',
+                'rows': 'N/A'
+            })
+
+        def removeInvalidClosedDates(table, report):
+            result = exec_sql(f"""
+                UPDATE {table}
+                SET closeddate = NULL
+                WHERE closeddate::timestamp < createddate::timestamp;
+            """)
+
+            report.append({
+                'description': 'removed invalid closed dates',
+                'rowsAffected': result.rowcount
+            })
+
+        log('\nCleaning ingest table.')
+        table = Ingest.__tablename__
+        report = []
+
+        dropDuplicates(table, report)
+        switchPrimaryKey(table, report)
+        removeInvalidClosedDates(table, report)
+
+        return report
+
+    async def populateDatabase(self,
+                               years=range(2015, 2021),
+                               limit=2000000,
+                               querySize=50000):
```

Comment:
We might want to move these default numbers to config that way theres a single source of truth

Reply:
I put the defaults in the config, but I consumed the defaults in app.py instead of here because those default values were also appearing in app.py. So the defaults get applied immediately if the query params are missing, and then passed into this service.